### ISSUE TITLE
[fast index] Deadlock found when trying to get lock

### ISSUE THREAD

**Issue by bato3** [maxbube/mydumper#342]
I'm using 24 threads to restore indexes, some tables are very big.  I got a couple of these errors (tables: `users` and `titles` are often referenced, some indexes can be restored over 20 mins): 
```
Error restoring m1040_strona.artifacts from file (null): Deadlock found when trying to get lock; try restarting transaction
ALTER TABLE `m1040_strona`.`artifacts`
 ADD  KEY `user_id` (`user_id`),
 ADD  KEY `title_id` (`title_id`),
 ADD  CONSTRAINT `artifacts_ibfk_1` FOREIGN KEY (`user_id`) REFERENCES `users` (`user_id`) ON DELETE SET NULL ON UPDATE CASCADE,
 ADD  CONSTRAINT `artifacts_ibfk_2` FOREIGN KEY (`title_id`) REFERENCES `titles` (`title_id`) ON DELETE CASCADE ON UPDATE CASCADE;
```
I suggest split `ALTER TABLE` to `ADD KEY` and `ADD  CONSTRAINT` or manage active locks.

### Info

`./myloader -d 20210512  -o -v 3 -t 24 -s m1040_strona --innodb-optimize-keys`

Dump from classic version (`mydumper 0.10.3, built against MySQL 5.7.33-36`), load from 10.05.2021 (`myloader 0.10.4, built against MySQL 8.0.22-13`, hash `1a2d1200d5108f480e9e8431cc52c2b0b1297059`)

https://github.com/maxbube/mydumper/pull/286

**Comment by davidducos** [comment]
Hi @bato3, 
Thank you for testing this new feature! I think that the issue is caused by the absence of the SET FOREIGN_KEY_CHEKS.
Please, check latest commit: https://github.com/maxbube/mydumper/pull/286/commits/15c07797bffa21ea9d8384df6659c17bf9d3ae91

As I think that it might fix the deadlock issue that you see. 

Please, let me know

**Comment by bato3** [comment]
Is better, but still don't work correctly. I noticed that another thread was inserting at the same time. 

Also, phpMyAdmin show in status column `waiting for table metadata lock` in all ALTER TABLE

**Comment by bato3** [comment]
@davidducos please reopen

**Comment by davidducos** [comment]
Hi @bato3, why don't you try with --sync-before-add-index ?

**Comment by bato3** [comment]
I'll check tomorrow. 

BTW: funniest thing is, that 2 firs errors are for empty tables, third for `innodb_online_alter_log_max_size`. 
But it's probably unrelated to that, because the last 3 errors are one index with a reference to the same table.

**Comment by bato3** [comment]
@davidducos still I have errors with `--sync-before-add-index` (the same tables)

Also I have only one process with `altering table` status. Others have `Waiting for table metadata lock`.

Blocking concurrent execution of `ALTER TABLE` is standard, or is it due to my database configuration?

If it's not global, I would like an option like `--alter-threads` or `--single-alter-table` to do this 1 thread per database. 


![image](https://user-images.githubusercontent.com/11570/118987653-c232e580-b980-11eb-891d-03901abdfcfc.png)

**Comment by davidducos** [comment]
@bato3 is the alter over the same table? if not, this doesn't make sense, as alters are being done in parallel and you should not see any "Waiting for table metadata lock", What db version are you running?

**Comment by bato3** [comment]
Yes, ALTER references the same tables.

Remember that in addition to `Waiting for table metadata lock`, I still get messages about` Deadlock found when trying to get lock; `and` innodb_online_alter_log_max_size`. I have set the value so that it only appears occasionally but never as I execute this query directly in the mysql console. 

DB Version:  8.0.22-13 - Percona Server (GPL), Release '13', Revision '6f7822f' (protocol 10)

```
# dpkg -l | grep -i percona
ii  libperconaserverclient21:amd64       8.0.22-13-1.focal                           amd64        Percona Server database client library
ii  libperconaserverclient21-dev         8.0.22-13-1.focal                           amd64        Percona Server database development files
ii  percona-release                      1.0-26.generic                              all          Package to install Percona gpg key and APT repos
ii  percona-server-client                8.0.22-13-1.focal                           amd64        Percona Server database client binaries
ii  percona-server-common                8.0.22-13-1.focal                           amd64        Percona Server database common files (e.g. /etc/mysql/my.cnf)
ii  percona-server-server                8.0.22-13-1.focal                           amd64        Percona Server database server binaries
ii  percona-server-source                8.0.22-13-1.focal                           amd64        Percona Server 8.0 source
ii  percona-toolkit                      3.3.1-1.focal                               amd64        Advanced MySQL and system command-line tools
ii  percona-xtrabackup-80                8.0.23-16-1.focal                           amd64        Open source backup tool for InnoDB and XtraDB
```

```
# mysql_config
Usage: /usr/bin/mysql_config [OPTIONS]
Compiler: GNU 9.3.0
Options:
        --cflags         [-I/usr/include/mysql ]
        --cxxflags       [-I/usr/include/mysql ]
        --include        [-I/usr/include/mysql]
        --libs           [-L/usr/lib/x86_64-linux-gnu -lperconaserverclient -lpthread -lm -lrt -lssl -lcrypto -ldl -lcoredumper -lresolv]
        --libs_r         [-L/usr/lib/x86_64-linux-gnu -lperconaserverclient -lpthread -lm -lrt -lssl -lcrypto -ldl -lcoredumper -lresolv]
        --plugindir      [/usr/lib/mysql/plugin]
        --socket         [/var/run/mysqld/mysqld.sock]
        --port           [0]
        --version        [8.0.22-13]
        --variable=VAR   VAR is one of:
                pkgincludedir [/usr/include/mysql]
                pkglibdir     [/usr/lib/x86_64-linux-gnu]
                plugindir     [/usr/lib/mysql/plugin]
```

```ini

[mysqld]
# === Required Settings ===
max_allowed_packet              = 256M
max_connect_errors              = 1000000
skip_external_locking
skip_name_resolve

# === SQL Compatibility Mode ===
sql_mode                       = ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION #,ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES

# === InnoDB Settings ===
default_storage_engine          = InnoDB
innodb_buffer_pool_instances    = 46     # Use 1 instance per 1GB of InnoDB pool size
innodb_buffer_pool_size         = 46G    # Use up to 70-80% of RAM
innodb_file_per_table           = 1
innodb_flush_log_at_trx_commit  = 0
innodb_flush_method             = O_DIRECT
innodb_log_buffer_size          = 16M
innodb_log_file_size            = 5G
innodb_stats_on_metadata        = 0

innodb_temp_data_file_path     = ibtmp1:64M:autoextend:max:20G # Control the maximum size for the ibtmp1 file
innodb_thread_concurrency      = 20    
innodb_read_io_threads          = 64
innodb_write_io_threads         = 64
#innodb_io_capacity             = 1000  # Max is 2000

key_buffer_size                 = 32M   # UPD

low_priority_updates            = 1
concurrent_insert               = 2

# === Connection Settings ===
max_connections                 = 250   

back_log                        = 512
thread_cache_size               = 100
thread_stack                    = 192K

interactive_timeout             = 180
wait_timeout                    = 180

# For MariaDB 10.1.1+ only (disabled by default)
#max_statement_time             = 30  

# === Buffer Settings ===
innodb_sort_buffer_size         = 2M    # UPD
join_buffer_size                = 4M    # UPD
read_buffer_size                = 3M    # UPD
read_rnd_buffer_size            = 4M    # UPD
sort_buffer_size                = 4M    # UPD

# === Table Settings ===
table_definition_cache          = 40000 # UPD
table_open_cache                = 40000 # UPD
open_files_limit                = 60000 
max_heap_table_size             = 128M
tmp_table_size                  = 128M

# === Search Settings ===
ft_min_word_len                 = 3     # Minimum length of words to be indexed for search results

# === Logging ===
log_error                       = /var/log/mysql/mysql_error.log
log_queries_not_using_indexes   = 1
long_query_time                 = 5
slow_query_log                  = 0     # Disabled for production
slow_query_log_file             = /var/log/mysql/mysql_slow.log

[mysqldump]

quote_names
max_allowed_packet              = 512M
transaction_isolation=read-committed
```

**Comment by davidducos** [comment]
@bato3,

ALTER on the same table can't be executed in parallel and take into account that myloader will add another ALTER statement if it finds FULLTEXT indexes, as it is possible to create 1 index per execution. 

I will need the DEADLOCK section from SHOW ENGINE INNODB STATUS to troubleshoot the deadlock issue.

**Comment by bato3** [comment]
I don't have `FULLTEXT` indexes. (but ~80% tables (~98% data amount) has references to 2 main tables)

At what point do I shoulld get the `SHOW ENGINE INNODB STATUS` result?  I need "step by step" help, because I don't have this section. After firs error I try:
```sql
SET GLOBAL innodb_print_all_deadlocks = 1;
 SHOW ENGINE INNODB STATUS;
```
Any messages in ` /var/log/mysql/mysql_error.log` only this section I'm think, that should be helpfull
```
TRANSACTIONS
------------
Trx id counter 1196567
Purge done for trx's n:o < 1196562 undo n:o < 0 state: running but idle
History list length 25
LIST OF TRANSACTIONS FOR EACH SESSION:
---TRANSACTION 422029522131256, not started
0 lock struct(s), heap size 1136, 0 row lock(s)
/*          --------- CUT COUPLE SIMILAR ROWS ------------  */
0 lock struct(s), heap size 1136, 0 row lock(s)
---TRANSACTION 1196563, ACTIVE 39 sec
mysql tables in use 1, locked 1
0 lock struct(s), heap size 1136, 0 row lock(s), undo log entries 3
MySQL thread id 309, OS thread handle 140554016061184, query id 152590 localhost root altering table
ALTER TABLE `m1040_strona`.`watched_chapters` 
 ADD  UNIQUE KEY `we_unique` (`user_id`,`chapter_id`),
 ADD  KEY `fk_watched_chapter1_idx` (`chapter_id`),
 ADD  KEY `fk_watched_chapters_manga1_idx` (`title_id`,`user_id`,`chapter_id`),
 ADD  CONSTRAINT `watched_chapters_ibfk_1` FOREIGN KEY (`user_id`) REFERENCES `users` (`user_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 ADD  CONSTRAINT `watched_chapters_ibfk_2` FOREIGN KEY (`chapter_id`) REFERENCES `chapters` (`chapter_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 ADD  CONSTRAINT `watched_chapters_ibfk_3` FOREIGN KEY (`title_id`) REFERENCES `mangas`
Trx read view will not see trx with id >= 1196567, sees < 1196567```

**Comment by davidducos** [comment]
@bato3 
There is a fix, that will be merged that might fix the issue that you have. Please check #351

**Comment by davidducos** [comment]
@bato3 it is rare to have deadlocks but do not see them in the Innodb status...

**Comment by bato3** [comment]
Nothing change to better, it's even worse... You didn't take into account how silly ideas database designers might have.


But back to the problem, I have another idea: The first 5 errors show up for tables that have no data. Only later do these errors appear for the data tables. (Also 5 errors) Could it be that the table lock was not released due to an error? 

> There is a fix, that will be merged that might fix the issue that you have. Please check #351

I have relation, who reference  to other column, not to primary key:

```
** Message: 13:09:56.004: Dropping table or view (if exists) `m1040_strona`.`dict_relations`
** Message: 13:09:56.074: Creating table `m1040_strona`.`dict_relations`
** Message: 13:09:56.075: Fast index creation will be use for table: m1040_strona.dict_relations

** (myloader:4052943): CRITICAL **: 13:09:56.122: Error restoring m1040_strona.dict_relations from file 
m1040_strona.dict_relations-schema.sql.gz: Failed to add the foreign key constraint. Missing index for constraint 
'protect_many2title_ibfk_3' in the referenced table 'dict_relations'
```
```sql
CREATE TABLE `dict_relations` (
  `dict_id` int(11) NOT NULL AUTO_INCREMENT,
  `dict_i18n_id` int(11) DEFAULT NULL,
  `lang` char(3) NOT NULL DEFAULT 'pl',
  `name` varchar(200) NOT NULL ,
  `relation_type` enum('staff2title','char2char','title2title','char2title','staff2char') NOT NULL,
  `relation_order` float NOT NULL DEFAULT '1000' ,
  PRIMARY KEY (`dict_id`)
) ENGINE=InnoDB AUTO_INCREMENT=191 DEFAULT CHARSET=utf8;

```
```sql
 CONSTRAINT `protect_many2title_ibfk_3` FOREIGN KEY (`staff_i18n_id`) 
        REFERENCES `dict_relations` (`dict_i18n_id`) ON DELETE CASCADE ON UPDATE CASCADE,
```

**Comment by davidducos** [comment]
Aha... but this has an easy fix, we currently:
- create the table with the primary/unique key.
- execute the inserts statements in parallel.
- create the indexes and constraints.
What about doing:
- create the table with the primary/unique key.
- execute the inserts statements in parallel.
- create the indexes
- create the constraints.
That will fix this issue.

**Comment by bato3** [comment]
> But back to the problem, I have another idea: The first 5 errors show up for tables that have no data. Only later do these errors appear for the data tables. (Also 5 errors) Could it be that the table lock was not released due to an error?

False alert. after add data to this rows - still have errors.

**Comment by davidducos** [comment]
@bato, can you check https://github.com/maxbube/mydumper/tree/fic_sync ? I think that this might help us to troubleshoot this issue

**Comment by bato3** [comment]
Hi @davidducos, 
It is much better. There are no longer **"Waiting for table metadata lock"** statuses, and there are fewer deadlocks. Though not quite...

When I was using the old database copy, there were only 5 deadlocks (10 before the patch). After refreshing, it increased to 8 locks. When I added additional messages to the console, I noticed 4 blockages. Well, normally the magic! XD
(Maybe it matters that I was pausing myloader (Ctrl + Z) during the last execution.)

I wonder if this could be related to the fact that query execution starts in the same millisecond ... But other places don't support my theory. 

And another magic: without logging in, there was **still a problem adding the `dict_relations` table**. But after my changes, he disappeared. 

I added an additional message at the end of thread execution (E), [after `restore_data_in_gstring`](https://github.com/maxbube/mydumper/blob/fic_sync/myloader.c#L612-L613) and immediately before executing the query (3) [before `mysql_real_query`](https://github.com/maxbube/mydumper/blob/fic_sync/myloader.c#L665-L666)

```
** Message: 20:25:29.498: Thread 14 restored (E) indexes or contraints `m1040_strona`.`watched_status`
** Message: 20:26:54.875: Thread 15 restored (E) indexes or contraints `m1040_strona`.`watched_statuslog`
** Message: 20:26:59.614: Thread 16 restored (E) indexes or contraints `m1040_strona`.`watched_chapters`
** Message: 20:57:31.664: Thread 19 restored (E) indexes or contraints `m1040_strona`.`rate_changelog`
** Message: 21:03:39.716: Thread 23 restored (E) indexes or contraints `m1040_strona`.`watched_episodes`
```
At this stage, it ended last long `ALTER TABLE` and began another series of operations. 
```
++ Message: 21:03:39.717: Thread 19 restoring (B) indexes or contraints `m1040_strona`.`pool_vote`
** Message: 21:03:39.718: Thread 23 restoring (B) indexes or contraints `m1040_strona`.`rate_titlecahracters`
>> Message: 21:03:39.718: Thread 15 restoring (B) indexes or contraints `m1040_strona`.`staff_versions`
++ Message: 21:03:39.718: Restoring query (3) m1040_strona.pool_vote from file (null):
>> Message: 21:03:39.718: Restoring query (3) m1040_strona.staff_versions from file (null):
** Message: 21:03:39.718: Thread 16 restoring (B) indexes or contraints `m1040_strona`.`rate_recommendations`
** Message: 21:03:39.718: Restoring query (3) m1040_strona.rate_titlecahracters from file (null):
** Message: 21:03:39.718: Thread 14 restoring (B) indexes or contraints `m1040_strona`.`rate_artifacts`
** Message: 21:03:39.718: Restoring query (3) m1040_strona.rate_recommendations from file (null):
** Message: 21:03:39.718: Restoring query (3) m1040_strona.rate_artifacts from file (null):
** Message: 21:03:39.718: Thread 11 restoring (B) indexes or contraints `m1040_strona`.`characters`
** Message: 21:03:39.718: Thread 1 restoring (B) indexes or contraints `m1040_strona`.`relation_title2title`
** Message: 21:03:39.718: Thread 3 restoring (B) indexes or contraints `m1040_strona`.`artifact_subtitles`
** Message: 21:03:39.719: Restoring query (3) m1040_strona.characters from file (null):
** Message: 21:03:39.718: Thread 8 restoring (B) indexes or contraints `m1040_strona`.`achievements`
** Message: 21:03:39.718: Thread 13 restoring (B) indexes or contraints `m1040_strona`.`chapters`
** Message: 21:03:39.719: Restoring query (3) m1040_strona.achievements from file (null):
** Message: 21:03:39.719: Restoring query (3) m1040_strona.artifact_subtitles from file (null):
** Message: 21:03:39.719: Restoring query (3) m1040_strona.relation_title2title from file (null):
** Message: 21:03:39.719: Restoring query (3) m1040_strona.chapters from file (null):
** Message: 21:03:39.719: Thread 4 restoring (B) indexes or contraints `m1040_strona`.`trait_filters`
** Message: 21:03:39.720: Restoring query (3) m1040_strona.trait_filters from file (null):
** Message: 21:03:39.719: Thread 5 restoring (B) indexes or contraints `m1040_strona`.`comment_artifacts`
** Message: 21:03:39.720: Restoring query (3) m1040_strona.comment_artifacts from file (null):
** Message: 21:03:39.719: Thread 21 restoring (B) indexes or contraints `m1040_strona`.`rate_stories`
** Message: 21:03:39.720: Thread 12 restoring (B) indexes or contraints `m1040_strona`.`rate_characters`
** Message: 21:03:39.719: Thread 7 restoring (B) indexes or contraints `m1040_strona`.`rate_reviews`
** Message: 21:03:39.721: Restoring query (3) m1040_strona.rate_stories from file (null):
** Message: 21:03:39.720: Thread 10 restoring (B) indexes or contraints `m1040_strona`.`club_versions`
** Message: 21:03:39.721: Restoring query (3) m1040_strona.rate_characters from file (null):
** Message: 21:03:39.721: Restoring query (3) m1040_strona.club_versions from file (null):
** Message: 21:03:39.720: Thread 17 restoring (B) indexes or contraints `m1040_strona`.`user_top`
** Message: 21:03:39.720: Thread 20 restoring (B) indexes or contraints `m1040_strona`.`pool_answer`
** Message: 21:03:39.721: Restoring query (3) m1040_strona.rate_reviews from file (null):
** Message: 21:03:39.721: Restoring query (3) m1040_strona.pool_answer from file (null):
** Message: 21:03:39.721: Restoring query (3) m1040_strona.user_top from file (null):
** Message: 21:03:39.720: Thread 18 restoring (B) indexes or contraints `m1040_strona`.`tag_descriptions`
** Message: 21:03:39.720: Thread 6 restoring (B) indexes or contraints `m1040_strona`.`comment_news`
** Message: 21:03:39.721: Thread 24 restoring (B) indexes or contraints `m1040_strona`.`club_artifacts`
** Message: 21:03:39.721: Restoring query (3) m1040_strona.tag_descriptions from file (null):
** Message: 21:03:39.721: Restoring query (3) m1040_strona.comment_news from file (null):
** Message: 21:03:39.721: Thread 22 restoring (B) indexes or contraints `m1040_strona`.`settings`
** Message: 21:03:39.721: Restoring query (3) m1040_strona.settings from file (null):
** Message: 21:03:39.721: Restoring query (3) m1040_strona.club_artifacts from file (null):
** Message: 21:03:39.721: Thread 9 restoring (B) indexes or contraints `m1040_strona`.`trait_descryption_versions`
** Message: 21:03:39.721: Thread 2 restoring (B) indexes or contraints `m1040_strona`.`relation_char2title`
** Message: 21:03:39.722: Restoring query (3) m1040_strona.trait_descryption_versions from file (null):
** Message: 21:03:39.722: Restoring query (3) m1040_strona.relation_char2title from file (null):
** Message: 21:03:39.729: Thread 8 restored (E) indexes or contraints `m1040_strona`.`achievements`
** Message: 21:03:39.730: Thread 8 restoring (B) indexes or contraints `m1040_strona`.`staff_names`
** Message: 21:03:39.730: Restoring query (3) m1040_strona.staff_names from file (null):
** Message: 21:03:39.733: Thread 6 restored (E) indexes or contraints `m1040_strona`.`comment_news`
** Message: 21:03:39.733: Thread 6 restoring (B) indexes or contraints `m1040_strona`.`staff_biography_versions`
** Message: 21:03:39.733: Thread 4 restored (E) indexes or contraints `m1040_strona`.`trait_filters`
** Message: 21:03:39.733: Thread 18 restored (E) indexes or contraints `m1040_strona`.`tag_descriptions`
** Message: 21:03:39.733: Restoring query (3) m1040_strona.staff_biography_versions from file (null):
** Message: 21:03:39.733: Thread 4 restoring (B) indexes or contraints `m1040_strona`.`episodes`
** Message: 21:03:39.733: Restoring query (3) m1040_strona.episodes from file (null):
** Message: 21:03:39.734: Thread 18 restoring (B) indexes or contraints `m1040_strona`.`relation_download2title`
** Message: 21:03:39.734: Restoring query (3) m1040_strona.relation_download2title from file (null):
** Message: 21:03:39.735: Thread 11 restored (E) indexes or contraints `m1040_strona`.`characters`
** Message: 21:03:39.735: Thread 11 restoring (B) indexes or contraints `m1040_strona`.`club_threads`
** Message: 21:03:39.735: Restoring query (3) m1040_strona.club_threads from file (null):
** Message: 21:03:39.741: Thread 23 restored (E) indexes or contraints `m1040_strona`.`rate_titlecahracters`
** Message: 21:03:39.742: Thread 23 restoring (B) indexes or contraints `m1040_strona`.`artifact_titles`
** Message: 21:03:39.742: Restoring query (3) m1040_strona.artifact_titles from file (null):
** Message: 21:03:39.751: Thread 13 restored (E) indexes or contraints `m1040_strona`.`chapters`
** Message: 21:03:39.751: Thread 13 restoring (B) indexes or contraints `m1040_strona`.`trait_descryptions`
** Message: 21:03:39.751: Restoring query (3) m1040_strona.trait_descryptions from file (null):
** Message: 21:03:39.755: Thread 16 restored (E) indexes or contraints `m1040_strona`.`rate_recommendations`
** Message: 21:03:39.755: Thread 16 restoring (B) indexes or contraints `m1040_strona`.`character_traits`
** Message: 21:03:39.755: Restoring query (3) m1040_strona.character_traits from file (null):
** Message: 21:03:39.762: Thread 1 restored (E) indexes or contraints `m1040_strona`.`relation_title2title`
** Message: 21:03:39.762: Thread 1 restoring (B) indexes or contraints `m1040_strona`.`rate_graphics`
** Message: 21:03:39.762: Restoring query (3) m1040_strona.rate_graphics from file (null):
++ Message: 21:03:39.768: Thread 19 restored (E) indexes or contraints `m1040_strona`.`pool_vote`
** Message: 21:03:39.768: Thread 19 restoring (B) indexes or contraints `m1040_strona`.`title_statuses`
** Message: 21:03:39.769: Restoring query (3) m1040_strona.title_statuses from file (null):
** Message: 21:03:39.781: Thread 10 restored (E) indexes or contraints `m1040_strona`.`club_versions`
** Message: 21:03:39.781: Thread 10 restoring (B) indexes or contraints `m1040_strona`.`callendar_events`
** Message: 21:03:39.781: Restoring query (3) m1040_strona.callendar_events from file (null):
** Message: 21:03:39.783: Thread 20 restored (E) indexes or contraints `m1040_strona`.`pool_answer`
** Message: 21:03:39.783: Thread 20 restoring (B) indexes or contraints `m1040_strona`.`rate_projects`
** Message: 21:03:39.783: Restoring query (3) m1040_strona.rate_projects from file (null):
** Message: 21:03:39.790: Thread 10 restored (E) indexes or contraints `m1040_strona`.`callendar_events`
** Message: 21:03:39.790: Thread 10 restoring (B) indexes or contraints `m1040_strona`.`user_purchases`
** Message: 21:03:39.790: Restoring query (3) m1040_strona.user_purchases from file (null):
** Message: 21:03:39.793: Thread 11 restored (E) indexes or contraints `m1040_strona`.`club_threads`
** Message: 21:03:39.793: Thread 11 restoring (B) indexes or contraints `m1040_strona`.`acl_user_sets`
** Message: 21:03:39.793: Restoring query (3) m1040_strona.acl_user_sets from file (null):
** Message: 21:03:39.794: Thread 7 restored (E) indexes or contraints `m1040_strona`.`rate_reviews`
** Message: 21:03:39.794: Thread 7 restoring (B) indexes or contraints `m1040_strona`.`title_stats`
** Message: 21:03:39.794: Restoring query (3) m1040_strona.title_stats from file (null):
** Message: 21:03:39.803: Thread 17 restored (E) indexes or contraints `m1040_strona`.`user_top`
** Message: 21:03:39.803: Thread 17 restoring (B) indexes or contraints `m1040_strona`.`tag_filters`
** Message: 21:03:39.803: Restoring query (3) m1040_strona.tag_filters from file (null):
** Message: 21:03:39.810: Thread 17 restored (E) indexes or contraints `m1040_strona`.`tag_filters`
** Message: 21:03:39.811: Thread 22 restored (E) indexes or contraints `m1040_strona`.`settings`
** Message: 21:03:39.810: Thread 17 restoring (B) indexes or contraints `m1040_strona`.`watched_status`
** Message: 21:03:39.812: Restoring query (3) m1040_strona.watched_status from file (null):
** Message: 21:03:39.812: Thread 22 restoring (B) indexes or contraints `m1040_strona`.`chapter_titles`
** Message: 21:03:39.812: Restoring query (3) m1040_strona.chapter_titles from file (null):
** Message: 21:03:39.826: Thread 8 restored (E) indexes or contraints `m1040_strona`.`staff_names`
** Message: 21:03:39.826: Thread 8 restoring (B) indexes or contraints `m1040_strona`.`character_biography_versions`
** Message: 21:03:39.827: Restoring query (3) m1040_strona.character_biography_versions from file (null):
** Message: 21:03:39.838: Thread 9 restored (E) indexes or contraints `m1040_strona`.`trait_descryption_versions`
** Message: 21:03:39.839: Thread 9 restoring (B) indexes or contraints `m1040_strona`.`ad_titles`
** Message: 21:03:39.839: Restoring query (3) m1040_strona.ad_titles from file (null):
** Message: 21:03:39.850: Thread 12 restored (E) indexes or contraints `m1040_strona`.`rate_characters`
** Message: 21:03:39.850: Thread 12 restoring (B) indexes or contraints `m1040_strona`.`comment_clubs`
** Message: 21:03:39.851: Restoring query (3) m1040_strona.comment_clubs from file (null):
** Message: 21:03:39.852: Thread 13 restored (E) indexes or contraints `m1040_strona`.`trait_descryptions`
** Message: 21:03:39.853: Thread 13 restoring (B) indexes or contraints `m1040_strona`.`title_points`
** Message: 21:03:39.853: Restoring query (3) m1040_strona.title_points from file (null):
** Message: 21:03:39.861: Thread 14 restored (E) indexes or contraints `m1040_strona`.`rate_artifacts`
** Message: 21:03:39.861: Thread 14 restoring (B) indexes or contraints `m1040_strona`.`club_like_characters`
** Message: 21:03:39.861: Restoring query (3) m1040_strona.club_like_characters from file (null):
** Message: 21:03:39.869: Thread 21 restored (E) indexes or contraints `m1040_strona`.`rate_stories`
** Message: 21:03:39.869: Thread 21 restoring (B) indexes or contraints `m1040_strona`.`title_threads`
** Message: 21:03:39.869: Restoring query (3) m1040_strona.title_threads from file (null):
** Message: 21:03:39.879: Thread 4 restored (E) indexes or contraints `m1040_strona`.`episodes`
** Message: 21:03:39.880: Thread 4 restoring (B) indexes or contraints `m1040_strona`.`subtitles`
** Message: 21:03:39.880: Restoring query (3) m1040_strona.subtitles from file (null):
** Message: 21:03:39.880: Thread 1 restored (E) indexes or contraints `m1040_strona`.`rate_graphics`
** Message: 21:03:39.880: Thread 1 restoring (B) indexes or contraints `m1040_strona`.`staff_points`
** Message: 21:03:39.880: Restoring query (3) m1040_strona.staff_points from file (null):
** Message: 21:03:39.890: Thread 18 restored (E) indexes or contraints `m1040_strona`.`relation_download2title`
** Message: 21:03:39.890: Thread 18 restoring (B) indexes or contraints `m1040_strona`.`chapter_online`
** Message: 21:03:39.890: Restoring query (3) m1040_strona.chapter_online from file (null):
** Message: 21:03:39.891: Thread 10 restored (E) indexes or contraints `m1040_strona`.`user_purchases`
** Message: 21:03:39.891: Thread 10 restoring (B) indexes or contraints `m1040_strona`.`dict_relations`
** Message: 21:03:39.892: Restoring query (3) m1040_strona.dict_relations from file (null):
** Message: 21:03:39.899: Thread 19 restored (E) indexes or contraints `m1040_strona`.`title_statuses`
** Message: 21:03:39.899: Thread 19 restoring (B) indexes or contraints `m1040_strona`.`character_versions`
** Message: 21:03:39.899: Restoring query (3) m1040_strona.character_versions from file (null):
** Message: 21:03:39.903: Thread 11 restored (E) indexes or contraints `m1040_strona`.`acl_user_sets`
** Message: 21:03:39.903: Thread 11 restoring (B) indexes or contraints `m1040_strona`.`episode_audiolang`
** Message: 21:03:39.903: Restoring query (3) m1040_strona.episode_audiolang from file (null):
** Message: 21:03:39.910: Thread 11 restored (E) indexes or contraints `m1040_strona`.`episode_audiolang`
** Message: 21:03:39.911: Thread 11 restoring (B) indexes or contraints `m1040_strona`.`title_news`
** Message: 21:03:39.911: Restoring query (3) m1040_strona.title_news from file (null):
** Message: 21:03:39.915: Thread 6 restored (E) indexes or contraints `m1040_strona`.`staff_biography_versions`
** Message: 21:03:39.915: Thread 6 restoring (B) indexes or contraints `m1040_strona`.`rate_designs`
** Message: 21:03:39.915: Restoring query (3) m1040_strona.rate_designs from file (null):
** Message: 21:03:39.930: Thread 20 restored (E) indexes or contraints `m1040_strona`.`rate_projects`
** Message: 21:03:39.930: Thread 20 restoring (B) indexes or contraints `m1040_strona`.`staffs`
** Message: 21:03:39.930: Restoring query (3) m1040_strona.staffs from file (null):
** Message: 21:03:39.942: Thread 12 restored (E) indexes or contraints `m1040_strona`.`comment_clubs`
** Message: 21:03:39.942: Thread 12 restoring (B) indexes or contraints `m1040_strona`.`artifact_characters`
** Message: 21:03:39.942: Restoring query (3) m1040_strona.artifact_characters from file (null):
** Message: 21:03:39.942: Thread 7 restored (E) indexes or contraints `m1040_strona`.`title_stats`
** Message: 21:03:39.942: Thread 7 restoring (B) indexes or contraints `m1040_strona`.`tag_description_versions`
** Message: 21:03:39.942: Restoring query (3) m1040_strona.tag_description_versions from file (null):
** Message: 21:03:39.943: Thread 1 restored (E) indexes or contraints `m1040_strona`.`staff_points`
** Message: 21:03:39.943: Thread 1 restoring (B) indexes or contraints `m1040_strona`.`watched_chapters`
** Message: 21:03:39.943: Restoring query (3) m1040_strona.watched_chapters from file (null):
** Message: 21:03:39.957: Thread 17 restored (E) indexes or contraints `m1040_strona`.`watched_status`
** Message: 21:03:39.957: Thread 17 restoring (B) indexes or contraints `m1040_strona`.`relation_char2char`
** Message: 21:03:39.957: Restoring query (3) m1040_strona.relation_char2char from file (null):
** Message: 21:03:39.970: Thread 22 restored (E) indexes or contraints `m1040_strona`.`chapter_titles`
** Message: 21:03:39.970: Thread 22 restoring (B) indexes or contraints `m1040_strona`.`staff_traits`
** Message: 21:03:39.970: Restoring query (3) m1040_strona.staff_traits from file (null):
** Message: 21:03:39.971: Thread 7 restored (E) indexes or contraints `m1040_strona`.`tag_description_versions`
** Message: 21:03:39.971: Thread 7 restoring (B) indexes or contraints `m1040_strona`.`anime_versions`
** Message: 21:03:39.971: Restoring query (3) m1040_strona.anime_versions from file (null):
** Message: 21:03:39.980: Thread 9 restored (E) indexes or contraints `m1040_strona`.`ad_titles`
** Message: 21:03:39.980: Thread 9 restoring (B) indexes or contraints `m1040_strona`.`episode_online`
** Message: 21:03:39.980: Restoring query (3) m1040_strona.episode_online from file (null):
** Message: 21:03:39.993: Thread 2 restored (E) indexes or contraints `m1040_strona`.`relation_char2title`
** Message: 21:03:39.993: Thread 2 restoring (B) indexes or contraints `m1040_strona`.`pool_question`
** Message: 21:03:39.993: Restoring query (3) m1040_strona.pool_question from file (null):
** Message: 21:03:39.995: Thread 9 restored (E) indexes or contraints `m1040_strona`.`episode_online`
** Message: 21:03:39.995: Thread 9 restoring (B) indexes or contraints `m1040_strona`.`projects`
** Message: 21:03:39.995: Restoring query (3) m1040_strona.projects from file (null):
** Message: 21:03:40.003: Thread 10 restored (E) indexes or contraints `m1040_strona`.`dict_relations`
** Message: 21:03:40.004: Thread 10 restoring (B) indexes or contraints `m1040_strona`.`club_like_tiltes`
** Message: 21:03:40.004: Restoring query (3) m1040_strona.club_like_tiltes from file (null):
** Message: 21:03:40.004: Thread 16 restored (E) indexes or contraints `m1040_strona`.`character_traits`
** Message: 21:03:40.004: Thread 16 restoring (B) indexes or contraints `m1040_strona`.`comments`
** Message: 21:03:40.004: Restoring query (3) m1040_strona.comments from file (null):
** Message: 21:03:40.006: Thread 2 restored (E) indexes or contraints `m1040_strona`.`pool_question`
** Message: 21:03:40.006: Thread 2 restoring (B) indexes or contraints `m1040_strona`.`recommendations`
** Message: 21:03:40.007: Restoring query (3) m1040_strona.recommendations from file (null):
** Message: 21:03:40.010: Thread 13 restored (E) indexes or contraints `m1040_strona`.`title_points`
** Message: 21:03:40.010: Thread 13 restoring (B) indexes or contraints `m1040_strona`.`mangas`
** Message: 21:03:40.010: Restoring query (3) m1040_strona.mangas from file (null):
** Message: 21:03:40.021: Thread 3 restored (E) indexes or contraints `m1040_strona`.`artifact_subtitles`
** Message: 21:03:40.021: Thread 3 restoring (B) indexes or contraints `m1040_strona`.`club_news`
** Message: 21:03:40.021: Restoring query (3) m1040_strona.club_news from file (null):

>>  (myloader:257834): CRITICAL **: 21:03:40.022: Error restoring m1040_strona.staff_versions from file (null): Deadlock found when trying to get lock; try restarting transaction
ALTER TABLE `m1040_strona`.`staff_versions`
 ADD  CONSTRAINT `fk_staff_versions_artifacts1` FOREIGN KEY (`picture_artifact_id`) REFERENCES `artifacts` (`artifact_id`) ON DELETE SET NULL ON UPDATE CASCADE,
 ADD  CONSTRAINT `fk_staff_versions_staffs1` FOREIGN KEY (`staff_id`) REFERENCES `staffs` (`staff_id`) ON DELETE SET NULL ON UPDATE CASCADE,
 ADD  CONSTRAINT `fk_staff_versions_users1` FOREIGN KEY (`user_id`) REFERENCES `users` (`user_id`) ON DELETE SET NULL ON UPDATE CASCADE;
>> Message: 21:03:40.022: Thread 15 restored (E) indexes or contraints `m1040_strona`.`staff_versions`
** Message: 21:03:40.022: Thread 15 restoring (B) indexes or contraints `m1040_strona`.`artifact_staffs`
** Message: 21:03:40.022: Restoring query (3) m1040_strona.artifact_staffs from file (null): 
```

**Comment by bato3** [comment]
Hi @davidducos 

1. It fails on `MyISAM` tables.
2. When I run without ` --innodb-optimize-keys` it also fails, but also on `InnoDB`
2. Error [with creation table](https://github.com/maxbube/mydumper/issues/342#issuecomment-849592202) "was fixed" after first `core dump`. Could the problem be that the keys are named the same as in another table? 



### add 1.

```
** Message: 09:08:18.070: Fast index creation will be use for table: m1040_strona.rate_lines
** Message: 09:08:18.097: Dropping table or view (if exists) `m1040_strona`.`traits`
** Message: 09:08:18.109: Creating table `m1040_strona`.`traits`
** Message: 09:08:18.109: Fast index creation will be use for table: m1040_strona.traits
** Message: 09:08:18.172: Thread 1 restoring `m1040_strona`.`rate_staffs` part 0 of 1. Progress 1 of 217 .
** Message: 09:08:18.173: Thread 2 restoring `m1040_strona`.`sb_tags` part 0 of 1. Progress 2 of 217 .
** Message: 09:08:18.173: Thread 3 restoring `m1040_strona`.`chapter_online` part 0 of 1. Progress 3 of 217 .
** Message: 09:08:18.175: Thread 4 restoring `m1040_strona`.`b8_categories` part 0 of 1. Progress 4 of 217 .
** Message: 09:08:18.176: Thread 4 restoring indexes `m1040_strona`.`b8_categories`
Segmentation fault (core dumped)
```
When I switched to 1 thread:
```
** Message: 09:25:23.182: 1 threads created
** Message: 09:25:23.182: Thread 1 restoring `m1040_strona`.`rate_staffs` part 0 of 1. Progress 1 of 217 .
** Message: 09:25:23.706: Thread 1 restoring indexes `m1040_strona`.`rate_staffs`
** Message: 09:25:23.943: Thread 1 restoring `m1040_strona`.`sb_tags` part 0 of 1. Progress 2 of 217 .
** Message: 09:25:23.951: Thread 1 restoring indexes `m1040_strona`.`sb_tags`
Segmentation fault (core dumped)

```


```sql
---  zcat m1040_strona.b8_categories*
/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;

/*!40103 SET TIME_ZONE='+00:00' */;
CREATE TABLE `b8_categories` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `category` varchar(4) DEFAULT NULL,
  `total` bigint(20) unsigned DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `category` (`category`)
) ENGINE=MyISAM AUTO_INCREMENT=3 DEFAULT CHARSET=utf8;


/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;
/*!40103 SET TIME_ZONE='+00:00' */;
INSERT INTO `b8_categories` VALUES
(1,"ham",0),
(2,"spam",0);
```

### add 2.
```
** Message: 09:51:15.182: Thread 1 restoring `m1040_strona`.`episodes` part 0 of 1. Progress 1 of 217 .
** Message: 09:51:15.182: Thread 2 restoring `m1040_strona`.`orient_titles` part 0 of 1. Progress 2 of 217 .
** Message: 09:51:15.183: Thread 3 restoring `m1040_strona`.`user_dbpass` part 0 of 1. Progress 3 of 217 .
** Message: 09:51:15.184: Thread 4 restoring `m1040_strona`.`mangas` part 0 of 1. Progress 4 of 217 .
** Message: 09:51:15.185: Thread 5 restoring `m1040_strona`.`redirections` part 0 of 1. Progress 5 of 217 .
** Message: 09:51:15.185: Thread 6 restoring `m1040_strona`.`title_urls` part 0 of 1. Progress 6 of 217 .
** Message: 09:51:15.186: Thread 7 restoring `m1040_strona`.`title_dmca` part 0 of 1. Progress 7 of 217 .
** Message: 09:51:15.187: Thread 8 restoring `m1040_strona`.`episode_online_cp` part 0 of 1. Progress 8 of 217 .
** Message: 09:51:15.188: Thread 9 restoring `m1040_strona`.`dict_relations` part 0 of 1. Progress 9 of 217 .
** Message: 09:51:15.189: Thread 10 restoring `m1040_strona`.`staff_names` part 0 of 1. Progress 10 of 217 .
** Message: 09:51:15.189: Thread 11 restoring `m1040_strona`.`watched_status` part 3 of 5. Progress 11 of 217 .
** Message: 09:51:15.190: Thread 12 restoring `m1040_strona`.`watched_status` part 1 of 5. Progress 12 of 217 .
** Message: 09:51:15.191: Thread 13 restoring `m1040_strona`.`watched_status` part 0 of 5. Progress 13 of 217 .
** Message: 09:51:15.193: Thread 14 restoring `m1040_strona`.`watched_status` part 4 of 5. Progress 14 of 217 .
** Message: 09:51:15.193: Thread 9 restoring indexes `m1040_strona`.`dict_relations`
Segmentation fault (core dumped)
```
```sql 
zcat m1040_strona.rate_staffs-schema.sql.gz
/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;

/*!40103 SET TIME_ZONE='+00:00' */;
CREATE TABLE `rate_staffs` (
  `rate_staff_id` int(11) NOT NULL AUTO_INCREMENT,
  `user_id` int(11) NOT NULL,
  `staff_id` int(11) NOT NULL,
  `rate` tinyint(4) DEFAULT NULL,
  `is_favourite` tinyint(1) NOT NULL DEFAULT '1',
  `rate_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `user_age` int(11) DEFAULT NULL,
  `fav_order` int(11) DEFAULT NULL,
  PRIMARY KEY (`rate_staff_id`),
  UNIQUE KEY `old_pk_rate_staffs` (`user_id`,`staff_id`),
  KEY `fk_users_has_staffs_staffs1_idx` (`staff_id`),
  CONSTRAINT `fk_users_has_staffs_staffs1` FOREIGN KEY (`staff_id`) REFERENCES `staffs` (`staff_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `fk_users_has_staffs_users1` FOREIGN KEY (`user_id`) REFERENCES `users` (`user_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=34342 DEFAULT CHARSET=utf8;
```

### add 3.
```sql

ALTER TABLE `protect_many2title`
  ADD PRIMARY KEY (`protect_id`),
  ADD UNIQUE KEY `unique` (`title_id`,`staff_id`,`staff_i18n_id`,`character_id`),
  ADD KEY `fk_relation_2title_titles1_idx` (`title_id`),
  ADD KEY `fk_relation_2title_staffs1_idx` (`staff_id`),
  ADD KEY `fk_relation_2title_dict_relations1_idx` (`staff_i18n_id`),
  ADD KEY `fk_relation_2title_characters1_idx` (`character_id`);

ALTER TABLE `relation_many2title`
  ADD PRIMARY KEY (`many_id`),
  ADD UNIQUE KEY `unique` (`title_id`,`staff_id`,`staff_i18n_id`,`character_id`,`character_i18n_id`),
  ADD KEY `fk_relation_2title_titles1_idx` (`title_id`),
  ADD KEY `fk_relation_2title_dict_relations1_idx` (`staff_i18n_id`),
  ADD KEY `fk_relation_2title_characters1_idx` (`character_id`),
  ADD KEY `fk_relation_2title_dict_relations2_idx` (`character_i18n_id`),
  ADD KEY `fk_relation_2title_staffs1_idx` (`staff_id`,`staff_i18n_id`);
```

**Comment by davidducos** [comment]
Hi @bato3 

1- We need to isolate this issue, I created a MyISAM table, and I was able to import without issue and no segmentation fault.
2- I was not able to reproduce this issue, are you testing latest version of [fic_sync](https://github.com/maxbube/mydumper/tree/fic_sync) ?
3- ALTER TABLE statement should NEVER add a PRIMARY KEY. if that is happening, it is a major issue. 

4- That is a new feature request, I will create the issue and see how to implement in the future... and it still surprising me... 

Yes! I' agree with you, I will implement a prettier  printing for the timings.

**Comment by davidducos** [comment]
Hi @bato3,
1 and 2- Please, check with the latest commit.
3- ALTER TABLE statement should NEVER add a PRIMARY KEY. if that is happening, it is a major issue.
4- That is a new feature request, I will create the issue and see how to implement in the future... and it still surprising me...

Yes! I' agree with you, I will implement a prettier printing for the timings.

**Comment by bato3** [comment]
**add 2.**, still fail when I use `./myloader -d /home/backups/site/20210609/ -p pass -o -v 3 -t 24 -s m1040_strona`. Myloader still try to restore indexes (both table types). And we're not using the delayed index addition method. `--innodb-optimize-keys`

**add 3.** Finally I understood what the problem was. Another table references a column that does not have an index. I'll post a new feature request for that. 

**add 4.** Overall, it would be nice to save failed queries to the file, so I'll post a separate issue for this. 

**add 1.** still have issues when import MyIsam tables. 

```sql
/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;

/*!40103 SET TIME_ZONE='+00:00' */;
CREATE TABLE `b8_categories` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `category` varchar(4) DEFAULT NULL,
  `total` bigint(20) unsigned DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `category` (`category`)
) ENGINE=MyISAM AUTO_INCREMENT=3 DEFAULT CHARSET=utf8;


/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;
/*!40103 SET TIME_ZONE='+00:00' */;
INSERT INTO `b8_categories` VALUES
(1,"ham",0),
(2,"spam",0);
```

**Comment by davidducos** [comment]
@bato3,
2- Are you testing https://github.com/maxbube/mydumper/tree/fic_sync ? if --innodb-optimize-keys is not used, you should not see any ALTER TABLE statement. Please, confirm this as it it quite strange. 

3- Please, share the table structure to reproduce the issue. 

4- Ok, thanks!

1- I had no issue with myisam tables with the latest fic_sync commit... can you check once more time and/or share the steps to reproduce it?

**Comment by bato3** [comment]
Hi,

Yes, `fic_sync` branch. This is pull from today:
![image](https://user-images.githubusercontent.com/11570/121536510-6bf71680-ca03-11eb-9ab7-b02592fc7e1f.png)
Anyway: I have to use it because I suggested changes in the summary of the times. **Thanks for that!** 



### Mysql version
 8.0.22-13 - Percona Server (GPL), Release '13', Revision '6f7822f' 
(Ubuntu 20... (LTS))

### 1 - crash on `restoring indexes`

for `./myloader -d /home/backups/site/20210609/ -p *** -o -v 3 -t 49 -s m1040_strona`
```
... prevous ...
** Message: 15:36:45.597: Creating table `m1040_strona`.`rate_musics`
** Message: 15:36:45.646: Dropping table or view (if exists) `m1040_strona`.`episode_online_cp`
** Message: 15:36:45.852: Creating table `m1040_strona`.`episode_online_cp`
** Message: 15:36:45.883: Dropping table or view (if exists) `m1040_strona`.`t_artifacts`
** Message: 15:36:45.896: Creating table `m1040_strona`.`t_artifacts`
** Message: 15:36:45.934: Dropping table or view (if exists) `m1040_strona`.`user_titles`
** Message: 15:36:45.956: Creating table `m1040_strona`.`user_titles`
** Message: 15:36:46.003: Dropping table or view (if exists) `m1040_strona`.`staff_biographies`
** Message: 15:36:46.017: Creating table `m1040_strona`.`staff_biographies`
** Message: 15:36:46.071: Dropping table or view (if exists) `m1040_strona`.`comment_titles`
** Message: 15:36:46.093: Creating table `m1040_strona`.`comment_titles`
** Message: 15:36:46.156: Dropping table or view (if exists) `m1040_strona`.`rate_lines`
** Message: 15:36:46.177: Creating table `m1040_strona`.`rate_lines`
** Message: 15:36:46.231: Dropping table or view (if exists) `m1040_strona`.`traits`
** Message: 15:36:46.248: Creating table `m1040_strona`.`traits`
** Message: 15:36:46.575: Thread 1 restoring `m1040_strona`.`orient_titles` part 0 of 1. Progress 1 of 211 .
** Message: 15:36:46.576: Thread 2 restoring `m1040_strona`.`user_dbpass` part 0 of 1. Progress 2 of 211 .
** Message: 15:36:46.577: Thread 3 restoring `m1040_strona`.`redirections` part 0 of 1. Progress 3 of 211 .
** Message: 15:36:46.578: Thread 4 restoring `m1040_strona`.`title_urls` part 0 of 1. Progress 4 of 211 .
** Message: 15:36:46.579: Thread 5 restoring `m1040_strona`.`title_dmca` part 0 of 1. Progress 5 of 211 .
** Message: 15:36:46.579: Thread 6 restoring `m1040_strona`.`episode_online_cp` part 0 of 1. Progress 6 of 211 .
** Message: 15:36:46.580: Thread 7 restoring `m1040_strona`.`staff_names` part 0 of 1. Progress 7 of 211 .
** Message: 15:36:46.581: Thread 8 restoring `m1040_strona`.`watched_status` part 3 of 5. Progress 8 of 211 .
** Message: 15:36:46.581: Thread 9 restoring `m1040_strona`.`watched_status` part 1 of 5. Progress 9 of 211 .
** Message: 15:36:46.582: Thread 10 restoring `m1040_strona`.`watched_status` part 0 of 5. Progress 10 of 211 .
** Message: 15:36:46.583: Thread 11 restoring `m1040_strona`.`watched_status` part 4 of 5. Progress 11 of 211 .
** Message: 15:36:46.583: Thread 12 restoring `m1040_strona`.`discord_messages` part 0 of 1. Progress 12 of 211 .
** Message: 15:36:46.584: Thread 13 restoring `m1040_strona`.`dict_relations` part 0 of 1. Progress 13 of 211 .
** Message: 15:36:46.585: Thread 14 restoring `m1040_strona`.`rate_lines` part 0 of 1. Progress 14 of 211 .
** Message: 15:36:46.585: Thread 15 restoring `m1040_strona`.`rate_titlecahracters` part 0 of 1. Progress 15 of 211 .
** Message: 15:36:46.586: Thread 16 restoring `m1040_strona`.`character_versions` part 0 of 1. Progress 16 of 211 .
** Message: 15:36:46.586: Thread 17 restoring `m1040_strona`.`staff_points` part 0 of 1. Progress 17 of 211 .
** Message: 15:36:46.587: Thread 18 restoring `m1040_strona`.`shop_items` part 0 of 1. Progress 18 of 211 .
** Message: 15:36:46.588: Thread 19 restoring `m1040_strona`.`artifacts` part 0 of 1. Progress 19 of 211 .
** Message: 15:36:46.589: Thread 20 restoring `m1040_strona`.`character_points` part 0 of 1. Progress 20 of 211 .
** Message: 15:36:46.589: Thread 13 restoring indexes `m1040_strona`.`dict_relations`
Segmentation fault (core dumped)
```

### 2 - crash on `MyISAM` 
for `./myloader -d /home/backups/site/myisa/ -p **** -o -v 3 -t 24 -s m1040_strona --innodb-optimize-keys` and 
for `./myloader -d /home/backups/site/myisa/ -p **** -o -v 3 -t 24 -s m1040_strona `
[structure](https://github.com/maxbube/mydumper/issues/342#issuecomment-857681942)
```
** Message: 15:51:56.160: Dropping table or view (if exists) `m1040_strona`.`b8_categories`
** Message: 15:51:56.167: Creating table `m1040_strona`.`b8_categories`
** Message: 15:51:56.174: Thread 1 restoring `m1040_strona`.`b8_categories` part 0 of 1. Progress 1 of 1 .
** Message: 15:51:56.175: Thread 1 restoring indexes `m1040_strona`.`b8_categories`
Segmentation fault (core dumped)
```

### 3 - Crash on bizarre constructions.  XD

There is a problem with the order of deletion. Because it first removes the `KEY` (along with the table) referenced by `CONSTRAINT` from the table to be removed later. 

Quick solution can be change order form `DELETE-CREATE-DELETE-CREATE` to `DELETE-DELETE-CREATE-CREATE`
```sql

CREATE TABLE `dict_relations` (
  `dict_id` int(11) NOT NULL AUTO_INCREMENT,
  `dict_i18n_id` int(11) DEFAULT NULL COMMENT 'if new - from atotoincrement id translated - from copy',
  `lang` char(3) NOT NULL DEFAULT 'pl',
  `name` varchar(200) NOT NULL ,
  `relation_type` enum('staff2title','char2char','title2title','char2title','staff2char') NOT NULL,
  `relation_order` float NOT NULL DEFAULT '1000' ,
  PRIMARY KEY (`dict_id`),
  UNIQUE KEY `uni1` (`dict_i18n_id`,`lang`),
  CONSTRAINT `fk_dict_relations_dict_relations1` FOREIGN KEY (`dict_i18n_id`) REFERENCES `dict_relations` (`dict_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=191 DEFAULT CHARSET=utf8;



CREATE TABLE `protect_many2title` (
  `protect_id` int(11) NOT NULL AUTO_INCREMENT,
  `title_id` int(11) NOT NULL,
  `staff_id` int(11) DEFAULT NULL,
  `staff_i18n_id` int(11) DEFAULT NULL,
  `character_id` int(11) DEFAULT NULL,
  `seiyuu_lang` char(3) DEFAULT NULL,
  PRIMARY KEY (`protect_id`),
  UNIQUE KEY `unique` (`title_id`,`staff_id`,`staff_i18n_id`,`character_id`,`seiyuu_lang`),
  KEY `fk_protect_2title_titles1_idx` (`title_id`) USING BTREE,
  KEY `fk_protect_2title_characters1_idx` (`character_id`) USING BTREE,
  KEY `fk_protect_2title_dict_relations1_idx` (`staff_i18n_id`) USING BTREE,
  KEY `fk_protect_2title_staffs1_idx` (`staff_id`) USING BTREE,
  CONSTRAINT `protect_many2title_ibfk_1` FOREIGN KEY (`title_id`) REFERENCES `titles` (`title_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `protect_many2title_ibfk_2` FOREIGN KEY (`staff_id`) REFERENCES `staffs` (`staff_id`) ON DELETE CASCADE ON UPDATE CASCADE,
-- ******************************
-- HERE. Uses `dict_i18n_id` not `dict_id`
-- ******************************
  CONSTRAINT `protect_many2title_ibfk_3` FOREIGN KEY (`staff_i18n_id`) REFERENCES `dict_relations` (`dict_i18n_id`) ON DELETE CASCADE ON UPDATE CASCADE,
-- ******************************
  CONSTRAINT `protect_many2title_ibfk_4` FOREIGN KEY (`character_id`) REFERENCES `characters` (`character_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=41607 DEFAULT CHARSET=utf8;
```

**Comment by davidducos** [comment]
Hi @bato3 

I'm testing with PS version 8.0.23-14.
1- I don't see any seg fault:
```bash
root@buster8:/tmp/mydumper# ./myloader -B fic_sync33 -d david -v 4 -o --innodb-optimize-keys
** Message: 14:38:19.521: Creating database `fic_sync33`
** Message: 14:38:19.522: Dropping table or view (if exists) `fic_sync33`.`protect_many2title`
** Message: 14:38:19.523: Creating table `fic_sync33`.`protect_many2title`
** Message: 14:38:19.524: Fast index creation will be use for table: fic_sync33.protect_many2title
** Message: 14:38:19.561: Dropping table or view (if exists) `fic_sync33`.`dict_relations`
** Message: 14:38:19.565: Creating table `fic_sync33`.`dict_relations`
** Message: 14:38:19.566: Fast index creation will be use for table: fic_sync33.dict_relations
** Message: 14:38:19.583: Thread 1 restoring indexes `fic_sync`.`protect_many2title`
** Message: 14:38:19.587: Thread 2 restoring indexes `fic_sync`.`dict_relations`
** Message: 14:38:19.591: 4 threads created
** Message: 14:38:19.609: Thread 1 restoring contraints `fic_sync`.`protect_many2title`
** Message: 14:38:19.609: Thread 3 shutting down
** Message: 14:38:19.609: Thread 2 restoring contraints `fic_sync`.`dict_relations`
** Message: 14:38:19.610: Thread 4 shutting down
** Message: 14:38:19.615: Thread 2 shutting down
** Message: 14:38:19.623: Thread 1 shutting down
** Message: 14:38:19.623: Import timings:
** Message: 14:38:19.623: Data      	| Index    	| Total   	| Table
** Message: 14:38:19.623: 0 00:00:00	| 0 00:00:00	| 0 00:00:00	| `fic_sync`.`protect_many2title`
** Message: 14:38:19.623: 0 00:00:00	| 0 00:00:00	| 0 00:00:00	| `fic_sync`.`dict_relations`
** Message: 14:38:19.623: Starting table checksum verification
```

3- this is the general log:
```bash
root@buster8:/tmp/mydumper# cat /var/lib/mysql/buster8.log
/usr/sbin/mysqld, Version: 8.0.23-14 (Percona Server (GPL), Release '14', Revision '3558242'). started with:
Tcp port: 3306  Unix socket: /var/run/mysqld/mysqld.sock
Time                 Id Command    Argument
2021-06-10T14:41:47.969320Z	   22 Quit
2021-06-10T14:41:50.046518Z	   23 Connect	root@localhost on  using Socket
2021-06-10T14:41:50.046826Z	   23 Query	SET SESSION wait_timeout = 2147483
2021-06-10T14:41:50.047062Z	   23 Query	SET SQL_LOG_BIN=0
2021-06-10T14:41:50.047264Z	   23 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.048072Z	   23 Query	SHOW CREATE DATABASE `fic_sync33`
2021-06-10T14:41:50.049796Z	   23 Query	DROP TABLE IF EXISTS `fic_sync33`.`protect_many2title`
2021-06-10T14:41:50.071323Z	   23 Query	DROP VIEW IF EXISTS `fic_sync33`.`protect_many2title`
2021-06-10T14:41:50.072383Z	   23 Query	USE `fic_sync33`
2021-06-10T14:41:50.072618Z	   23 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.072740Z	   23 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.072814Z	   23 Query	/*!40103 SET TIME_ZONE='+00:00' */
2021-06-10T14:41:50.073396Z	   23 Query	CREATE TABLE `protect_many2title` (
  `protect_id` int NOT NULL AUTO_INCREMENT,
  `title_id` int NOT NULL,
  `staff_id` int DEFAULT NULL,
  `staff_i18n_id` int DEFAULT NULL,
  `character_id` int DEFAULT NULL,
  `seiyuu_lang` char(3) DEFAULT NULL,
  PRIMARY KEY (`protect_id`)
) ENGINE=InnoDB AUTO_INCREMENT=41607 DEFAULT CHARSET=utf8
2021-06-10T14:41:50.085926Z	   23 Query	SHOW CREATE DATABASE `fic_sync33`
2021-06-10T14:41:50.086686Z	   23 Query	DROP TABLE IF EXISTS `fic_sync33`.`dict_relations`
2021-06-10T14:41:50.094703Z	   23 Query	DROP VIEW IF EXISTS `fic_sync33`.`dict_relations`
2021-06-10T14:41:50.095786Z	   23 Query	USE `fic_sync33`
2021-06-10T14:41:50.095935Z	   23 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.096045Z	   23 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.096207Z	   23 Query	/*!40103 SET TIME_ZONE='+00:00' */
2021-06-10T14:41:50.096725Z	   23 Query	CREATE TABLE `dict_relations` (
  `dict_id` int NOT NULL AUTO_INCREMENT,
  `dict_i18n_id` int DEFAULT NULL COMMENT 'if new - from atotoincrement id translated - from copy',
  `lang` char(3) NOT NULL DEFAULT 'pl',
  `name` varchar(200) NOT NULL,
  `relation_type` enum('staff2title','char2char','title2title','char2title','staff2char') NOT NULL,
  `relation_order` float NOT NULL DEFAULT '1000',
  PRIMARY KEY (`dict_id`)
) ENGINE=InnoDB AUTO_INCREMENT=191 DEFAULT CHARSET=utf8
2021-06-10T14:41:50.110405Z	   24 Connect	root@localhost on  using Socket
2021-06-10T14:41:50.110588Z	   24 Query	SET SESSION wait_timeout = 2147483
2021-06-10T14:41:50.110735Z	   24 Query	SET SQL_LOG_BIN=0
2021-06-10T14:41:50.110883Z	   24 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.111078Z	   24 Query	/*!40101 SET SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */
2021-06-10T14:41:50.111219Z	   24 Query	/*!40014 SET UNIQUE_CHECKS=0 */
2021-06-10T14:41:50.111370Z	   24 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.111507Z	   24 Query	SET autocommit=0
2021-06-10T14:41:50.112200Z	   24 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.112412Z	   24 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.112724Z	   24 Query	/*!40103 SET TIME_ZONE='+00:00' */
2021-06-10T14:41:50.112917Z	   24 Query	ALTER TABLE `fic_sync33`.`protect_many2title`
 ADD  UNIQUE KEY `unique` (`title_id`,`staff_id`,`staff_i18n_id`,`character_id`,`seiyuu_lang`),
 ADD  KEY `fk_protect_2title_titles1_idx` (`title_id`) USING BTREE,
 ADD  KEY `fk_protect_2title_characters1_idx` (`character_id`) USING BTREE,
 ADD  KEY `fk_protect_2title_dict_relations1_idx` (`staff_i18n_id`) USING BTREE,
 ADD  KEY `fk_protect_2title_staffs1_idx` (`staff_id`) USING BTREE
2021-06-10T14:41:50.115684Z	   25 Connect	root@localhost on  using Socket
2021-06-10T14:41:50.115850Z	   25 Query	SET SESSION wait_timeout = 2147483
2021-06-10T14:41:50.116009Z	   25 Query	SET SQL_LOG_BIN=0
2021-06-10T14:41:50.116080Z	   25 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.116234Z	   25 Query	/*!40101 SET SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */
2021-06-10T14:41:50.116300Z	   25 Query	/*!40014 SET UNIQUE_CHECKS=0 */
2021-06-10T14:41:50.116342Z	   25 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.116427Z	   25 Query	SET autocommit=0
2021-06-10T14:41:50.117797Z	   26 Connect	root@localhost on  using Socket
2021-06-10T14:41:50.118098Z	   26 Query	SET SESSION wait_timeout = 2147483
2021-06-10T14:41:50.118196Z	   25 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.118247Z	   26 Query	SET SQL_LOG_BIN=0
2021-06-10T14:41:50.118303Z	   25 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.118345Z	   25 Query	/*!40103 SET TIME_ZONE='+00:00' */
2021-06-10T14:41:50.118434Z	   26 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.118662Z	   26 Query	/*!40101 SET SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */
2021-06-10T14:41:50.118856Z	   26 Query	/*!40014 SET UNIQUE_CHECKS=0 */
2021-06-10T14:41:50.119025Z	   26 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.119224Z	   25 Query	ALTER TABLE `fic_sync33`.`dict_relations`
 ADD  UNIQUE KEY `uni1` (`dict_i18n_id`,`lang`)
2021-06-10T14:41:50.121167Z	   26 Query	SET autocommit=0
2021-06-10T14:41:50.122182Z	   27 Connect	root@localhost on  using Socket
2021-06-10T14:41:50.122377Z	   27 Query	SET SESSION wait_timeout = 2147483
2021-06-10T14:41:50.122565Z	   27 Query	SET SQL_LOG_BIN=0
2021-06-10T14:41:50.122650Z	   27 Query	/*!40101 SET NAMES binary*/
2021-06-10T14:41:50.122781Z	   27 Query	/*!40101 SET SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */
2021-06-10T14:41:50.123017Z	   27 Query	/*!40014 SET UNIQUE_CHECKS=0 */
2021-06-10T14:41:50.123129Z	   27 Query	/*!40014 SET FOREIGN_KEY_CHECKS=0*/
2021-06-10T14:41:50.123167Z	   27 Query	SET autocommit=0
2021-06-10T14:41:50.137586Z	   27 Quit
2021-06-10T14:41:50.137773Z	   24 Quit
2021-06-10T14:41:50.137789Z	   26 Query	ALTER TABLE `fic_sync33`.`dict_relations`
 ADD  CONSTRAINT `fk_dict_relations_dict_relations1` FOREIGN KEY (`dict_i18n_id`) REFERENCES `dict_relations` (`dict_id`) ON DELETE CASCADE ON UPDATE CASCADE
2021-06-10T14:41:50.142188Z	   25 Query	ALTER TABLE `fic_sync33`.`protect_many2title`
 ADD  CONSTRAINT `protect_many2title_ibfk_1` FOREIGN KEY (`title_id`) REFERENCES `titles` (`title_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 ADD  CONSTRAINT `protect_many2title_ibfk_2` FOREIGN KEY (`staff_id`) REFERENCES `staffs` (`staff_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 ADD  CONSTRAINT `protect_many2title_ibfk_3` FOREIGN KEY (`staff_i18n_id`) REFERENCES `dict_relations` (`dict_i18n_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 ADD  CONSTRAINT `protect_many2title_ibfk_4` FOREIGN KEY (`character_id`) REFERENCES `characters` (`character_id`) ON DELETE CASCADE ON UPDATE CASCADE
2021-06-10T14:41:50.143468Z	   26 Quit
2021-06-10T14:41:50.151329Z	   23 Quit
2021-06-10T14:41:50.151850Z	   25 Quit
2021-06-10T14:41:52.543733Z	   28 Connect	root@localhost on  using Socket
2021-06-10T14:41:52.545293Z	   28 Query	select @@version_comment limit 1
2021-06-10T14:41:55.538285Z	   28 Query	set global general_log=OFF
```
The SET FOREIGN_KEY_CHECKS=0 is there all the time. The scenario that you are presenting is this?
```mysql
mysql> DROP TABLE IF EXISTS `fic_sync33`.`dict_relations`;
ERROR 3730 (HY000): Cannot drop table 'dict_relations' referenced by a foreign key constraint 'protect_many2title_ibfk_3' on table 'protect_many2title'.
```
which is avoided using foreign_key_checks off:
```mysql
mysql> set foreign_key_checks= OFF;
Query OK, 0 rows affected (0.00 sec)

mysql> DROP TABLE IF EXISTS `fic_sync33`.`dict_relations`;
Query OK, 0 rows affected (0.04 sec)
```

**Comment by bato3** [comment]
### deadlocks
I don't know what causes them. With 24 threads (1 / core), the problem occurs. As I ran with 4 threads - no problem.

Can you verify that all necessary queries are performed when restoring constants? I remember, that you add `SET FOREIGN_KEY_CHECKS = 0` before merging the previous branch. Maybe something is missing after separating the `KEY`s from the `CONSTRAINT`s.

**And due to the fact that the most time-consuming key restoration operation has been separated, it may be worth restoring constants in 1 thread. (As is the case when creating tables).** 

### 1 & 2 - can confirm, fixed after last commit

### 3 - I am sending you a complete example. 
 1. db structure:
```sql
CREATE TABLE `ref_table` (
  `id` int NOT NULL AUTO_INCREMENT,
  `strange_ref` int NOT NULL,
  PRIMARY KEY (`id`),
  KEY `strange_ref` (`strange_ref`),
  CONSTRAINT `ref_table_ibfk_1` FOREIGN KEY (`strange_ref`) REFERENCES `strange_table` (`id_copy`) ON DELETE RESTRICT ON UPDATE RESTRICT
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;


CREATE TABLE `strange_table` (
  `id` int NOT NULL AUTO_INCREMENT,
  `id_copy` int NOT NULL,
  PRIMARY KEY (`id`),
  KEY `id_copy` (`id_copy`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```
2. run import to clean database
3. restart import (no cleanup) 

**But the most important thing is the order in which `myloader` performs operations.** 
The error occurs if there is `strange_table` first, then `ref_table`. 
I tried to get this error on `mysqld  Ver 8.0.25 for Linux on x86_64 (MySQL Community Server - GPL)`, but I can't change the order in which `myloader` executes queries.

**Comment by davidducos** [comment]
@bato3 v0.10.7 will be released next week. I will try to push https://github.com/maxbube/mydumper/tree/fic_sync and merge a couple of weeks after, so til that point we are going to leave this on hold. 
Sorry, but I'm not able to reproduce the issue. However, after merge, we are going to be able to change the order of the tables. Please, be patient and hang in there!

**Comment by bato3** [comment]
@davidducos 

Thanks for your work!
With the already implemented changes, it is almost perfectly.


The table restore order error is so specific that I don't know if 1% of users will encounter it. So you can ignore it. Alternatively, `3730` will display a prompt:" The foreign key is referencing a column that is not PRIMARY_KEY. We recommend importing into an empty database. "

**As for restoring constants** (deadlocks error): There is a chance to do this in 1 thread? It takes little time, and this form would solve the problem. If not, I suggest you close the application with tags like: `need more info`, `on hold`

**Comment by davidducos** [comment]
does `restoring constants` means `restoring constraints`? if not, I'm not sure what do you mean...

**Comment by bato3** [comment]
Sorry for my _Machine Translated_ English ;)

I'm talk about eg.: 
```sql
ALTER TABLE `fic_sync33`.`dict_relations`
 ADD  CONSTRAINT `fk_dict_relations_dict_relations1` FOREIGN KEY (`dict_i18n_id`) REFERENCES `dict_relations` (`dict_id`) ON DELETE CASCADE ON UPDATE CASCADE
```
